docs: updated and added OIDC credential forwarding  COPS-264, COPS-284, COPS-319 (#2692)
neo4j/docs-operations: modules/ROOT/pages/database-administration/aliases/remote-database-alias-configuration.adoc

[[manage-remote-aliases]]
= Configuring remote database aliases

You can use a remote database alias to connect to one or more graphs or to one or more remote standalone servers or clusters.

Although remote database aliases do not store any data, they enable users or applications to perform queries on remote databases as if they were on the local DBMS server.
All configurations can be done using xref:database-administration/aliases/manage-aliases-standard-databases.adoc[administrative commands] on a running system.
Any changes are automatically synchronized across all members of a cluster.

The following steps describe the setup required to define a remote database alias for both local and remote DBMSs.

> They assume that you have already installed and started both DBMSs: a local *DBMS A* and a remote *DBMS B*, and that you have administrator access to both systems.

Reviewer: ah, ok then. If this arrow is enough to show local and remote DBMS, let's do not overload the image.
I think I was puzzled, because in the example, the first step is about the remote dbms and the second step is about the local dbms.

* `STORED NATIVE CREDENTIALS`, the credentials of a single native user on the remote *DBMS B*.
* label:new[Introduced in 2025.x] `OIDC CREDENTIAL FORWARDING`, forwarding the bearer authentication token from the logged-in user on the local *DBMS A*.
The user needs to be logged in with an identity provider supporting OIDC.

[[setup-example-stored-native-credentials]]
== Setup example with stored native credentials

In this example, _Alice_ is an administrator and _Carol_ is a user who needs access to a database managed by _Bob_.

image::remote-alias-overview.svg[title="Overview of the required remote database alias setup when using stored credentials", role="middle"]

A remote database alias is only accessible to users with appropriate privileges.
In this example, _Bob_ is the administrator responsible for deciding which database (`db1` or `db2`) the remote aliases can write and/or read.
Meanwhile, _Alice_ is the administrator that assigns who has access to the privileges set by _Bob_.
In the example, _Alice_ will assign that access to _Carol_.
See xref:authentication-authorization/dbms-administration.adoc[DBMS privileges] for more information.

> _Carol_ can use her own regular credentials to access the remote database `db1` in *DBMS B* after _Alice_ assigns this privilege to her user profile.

Reviewer: If Carol can access db1 directly, why would she need the access to the remote alias of this db1?

Then, _Alice_ will be able to grant _Carol_ access to the remote database alias `db1-remote-alias` that connects to `db1` on *DBMS B* using the shared credentials.
//This configuration will also allow _Carol_ to access `db2` in *DBMS B*, if _Bob_ grants the necessary privileges to the user profile shared with _Alice_.
//See xref:authentication-authorization/index.adoc[Authentication and authorization] for further information.

=== Configure the remote DBMS B (_Bob_)

As _Bob_, you are responsible for the remote *DBMS B*.
You can create and delete users and grant or deny privileges on the databases managed by *DBMS B*.

In this example, you create a user called `alias_user`, which will be used by the remote database alias to connect to `db1`, and shares the credentials with _Alice_.

. Create the user profile to share with _Alice_:
+
[source, Cypher]
----
CREATE USER alias_user SET PASSWORD 'secretpassword'
----
. Create a custom role to track all users shared on a remote connection, so that they remain trackable:
+
[source, Cypher]
----
CREATE ROLE remote_access
----

> . Grant the `remote_access` role access to `db1` and assign the role to the user profile created for the remote database alias, `alias_user`:

Reviewer: ```suggestion
. Grant the `remote_access` role to `db1` and assign the role to the user profile created for the remote database alias, `alias_user`:
```

Where `password.sh` might look like this:

.Password.sh
[source, bash]
----
#!/bin/bash
echo "$KEYSTORE_PASSWORD_ENVIRONMENT_VARIABLE"
----

Additionally, do not forget to change the permissions of the configuration file and start Neo4j with the command expansion flag:

[source, bash]
----
chmod 640 conf/neo4j.conf
bin/neo4j start --expand-commands
----

> === Manage remote database aliases

Reviewer: This section is about  Carol, isn’t it? When she gets all privileges, she can manage the remote database alias

Reviewer 2: I think so.

This needs to be set by an administrator.
See xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-user-management[User Management] for more information.
+
[source, Cypher]
----
ALTER USER alice SET HOME DATABASE `db1-remote-alias`
----

== Important notes

When using remote database aliases, keep in mind that:

* Remote alias transactions will not be visible in `SHOW TRANSACTIONS` on *DBMS A*.
However, they can be accessed and terminated on the remote database when connecting with the same user.

* Actions on the remote *DBMS B* are all attributed to the user configured for the remote database alias.
In the case of using `STORED NATIVE CREDENTIALS`, the same credentials are used to connect to the remote DBMS regardless of which end-user made the query targeting the remote alias.
This will result in the stored native user being logged in the audit trails on the remote DBMS for all queries using the remote database alias.
When using `OIDC CREDENTIAL FORWARDING`, the actual end-user's credentials and permissions are used, resulting in per-user audit trails being logged on the remote DBMS.

> * When using a remote database alias with OIDC credential forwarding, the user needs to be logged in to the local DBMS with OIDC, otherwise there is no token to forward, and the access to the remote database will be denied with GQLSTATUS link:https://neo4j.com/docs/status-codes/current/errors/gql-errors/42NFF/[`42NFF`].

Reviewer: Maybe, I didn't fully understand the concept of OIDC credential forwarding, but here a local DBMS is mentioned, and there is no local DBMS on the figure 2.

Reviewer: I would say that it will clarify the situation, but Reneta mentioned earlier that the arrow 'remote connection' implies that the DBMS A is local and DBMS B is remote.

This configuration independence can lead to privilege inconsistency (e.g., over-privileging or unexpected access denial).
====

=== Configure a remote DBMS (_Bob_)

In the suggested example, there are two administrators: _Alice_ and _Bob_.
_Bob_ is the administrator responsible for the setup of the remote DBMS, which includes the following steps:

. Set up SSO and support for the identity provider.
. Map the identity provider groups to the Neo4j roles. If you don’t want specific users to access `Db2`, here is where you set it.

Additionally, _Bob_ must do the required setup for the link:https://neo4j.com/docs/operations-manual/current/security/ssl-framework/[SSL framework] and check whether the database accepts a non-local connection if required.

[source, Example of additional configuration]
----
# accept non-local connections
server.default_listen_address=0.0.0.0

# configure ssl for bolt
dbms.ssl.policy.bolt.enabled=true
dbms.ssl.policy.bolt.base_directory=certificates/bolt
dbms.ssl.policy.bolt.private_key=private.key
dbms.ssl.policy.bolt.public_certificate=public.crt
dbms.ssl.policy.bolt.client_auth=NONE

# enforcing ssl connection

> server.bolt.tls_level=REQUIRED

Author: Went for the second option here

[source, bash]
----
chmod 640 conf/neo4j.conf
bin/neo4j start --expand-commands
----

==== Create the remote database alias and grant access to Carol

You create the remote database alias using xref:database-administration/aliases/manage-aliases-standard-databases.adoc[alias administrative commands] and grant _Carol_ access to it.

[NOTE]
====
It is strongly recommended to connect to a remote database alias with a secured connection.
Note that only client-side SSL is supported.
By default, remote database aliases require a secured URI scheme such as `neo4j+s`.
However, if you want to disable the secure URL scheme, you can set the driver setting `ssl_enforced` to `false`.
====

. Use the following command to create a remote database alias with the stored native credentials shared by _Bob_:
+
[source, Cypher]
----
CREATE ALIAS `db1-remote-alias` FOR DATABASE `db1` AT "neo4j+s://location:7687" USER remote_user PASSWORD 'secretpassword'
----

.Create a custom role to track all users who have access to the remote database alias (or chose an already existing role):

> [source, Cypher]

Reviewer 2: ```suggestion
+
[source, Cypher]
```

You set up SSO on the local *DBMS A* and map the identity provider groups to the Neo4j roles.
For details, see the xref:tutorial/tutorial-sso-configuration.adoc[SSO configuration tutorial] and xref:authentication-authorization/sso-integration.adoc#auth-sso-map-idp-roles[Map the identity provider groups to the Neo4j roles].

[parameters]
----
dbms.security.oidc.<provider>.well_known_discovery_uri=http://example.com/.well-known/discovery
<...>
dbms.security.oidc.<provider>.claims.groups=groups
dbms.security.oidc.<provider>.authorization.group_to_role_mapping= "engineers" = admin; \
                                                                   "collaborators" = reader; \
                                                                   "remote_users" = remote_access
----

=== Configure the remote DBMS B (_Bob_)

As _Bob_, you are responsible for setting up the remote *DBMS B*.
You can create and delete users and grant or deny privileges on the databases managed by *DBMS B*.

In this example, you need to ensure that _Carol_ can access `db1` on *DBMS B* using OIDC credential forwarding.

> . Create a custom role or chose an existing one, the roles will be assigned to the users on log in via the mapping of identity provider groups to the Neo4j roles:

Reviewer 2: ```suggestion
. Create a custom role or choose an existing one.
The roles will be assigned to the users on login via the mapping of identity provider groups to the Neo4j roles:
```

+
[source, Cypher]
----
CREATE ROLE db1_access
----

> . Grant the `db1_access` role access to `db1`.

Reviewer 2: ```suggestion
. Grant the `db1_access` role access to `db1`:
```

+
[source, Cypher]
----
GRANT ACCESS ON DATABASE db1 TO db1_access
GRANT MATCH {*} ON GRAPH db1 TO db1_access
----

. Set up SSO on the remote *DBMS B* and map the identity provider groups to the Neo4j roles.
For details, see the xref:tutorial/tutorial-sso-configuration.adoc[SSO configuration tutorial] and xref:authentication-authorization/sso-integration.adoc#auth-sso-map-idp-roles[Map the identity provider groups to the Neo4j roles].
//If you do not want specific users to access `db2`, here is where you set it.
+
[parameters]
----
dbms.security.oidc.<provider>.well_known_discovery_uri=http://example.com/.well-known/discovery
<...>
dbms.security.oidc.<provider>.claims.groups=groups
dbms.security.oidc.<provider>.authorization.group_to_role_mapping= "engineers" = admin; \
                                                                   "collaborators" = reader; \

>                                                                    "remote_users" = db1_access

Reviewer 2: Here, the role is different from the one configured on DBMS A. Don't they need to be the same?

[CAUTION]
====
While it is permitted to use different OIDC configurations across distinct DBMS instances (e.g., local vs. target), users must be aware of the resulting privilege disparity.
A user's effective permissions are not dictated by the identity provider groups alone, but by the mapping of those groups to the roles defined within each specific Neo4j DBMS, see xref:authentication-authorization/sso-integration.adoc#auth-sso-map-idp-roles[Map the identity provider groups to the Neo4j roles].
Crucially, if the OIDC configuration settings differ between the local DBMS and the target DBMS, the user will have different effective privileges on those systems.
This configuration independence can lead to privilege inconsistency (e.g., over-privileging or unexpected access denial).
====

=== Configure a remote DBMS (_Bob_)

In the suggested example, there are two administrators: _Alice_ and _Bob_.
_Bob_ is the administrator responsible for the setup of the remote DBMS, which includes the following steps:

. Set up SSO and support for the identity provider.

> . Map the identity provider groups to the Neo4j roles. If you don’t want specific users to access `Db2`, here is where you set it.

Reviewer 3: Maybe this is a good point to link to
https://neo4j.com/docs/operations-manual/current/tutorial/tutorial-sso-configuration/

Same for "Configure local DBMS" below

====
It is strongly recommended to connect to a remote database alias with a secured connection.
Note that only client-side SSL is supported.
By default, remote database aliases require a secured URI scheme such as `neo4j+s`.
However, if you want to disable the secure URL scheme, you can set the driver setting `ssl_enforced` to `false`.
====

. Use the following command to create a remote database alias with the stored native credentials shared by _Bob_:
+
[source, Cypher]
----
CREATE ALIAS `db1-remote-alias` FOR DATABASE `db1` AT "neo4j+s://location:7687" USER remote_user PASSWORD 'secretpassword'
----

.Create a custom role to track all users who have access to the remote database alias (or chose an already existing role):

[source, Cypher]
----

> CREATE ROLE remote_access

Author: The remote_access role needs to be created before it is granted to Carol.

==== Changing the encryption key

> If the encryption key in the keystore is changed, the encrypted credentials for all existing remote database aliases requires updating, as they will no longer be readable with the new key.

Author: Since this is not a step in the process of setting it up, but one that needs to be taken if the keystore is changed, I changed the wording a bit here.
